### PR TITLE
Move RingBuffer to a separate module

### PR DESCRIPTION
**Pull request by mgeier**
This uses the freshly created module `pa_ringbuffer`: https://github.com/mgeier/python-pa-ringbuffer.

Thanks to @arigo for helping me on the CFFI mailing list: https://groups.google.com/forum/#!topic/python-cffi/JtAKU-g9Exg!

@tgarc Would you mind testing if that works for you?

Any comments?
Suggestions for improvements?

### CODE REVIEW

**Comment by tgarc** [comment]
Yes, this module seems to work fine for me. Thanks!

**Comment by mgeier** [comment]
@tgarc Thanks for testing!
Which OS did you try?

I initially tried it on Debian Linux and there it works.

Sadly, it doesn't seem to work on macOS:

```
$ python3 setup.py develop
...
creating build/temp.macosx-10.11-x86_64-3.6/build/temp.macosx-10.11-x86_64-3.6
clang -Wno-unused-result -Wsign-compare -Wunreachable-code -fno-common -dynamic -DNDEBUG -g -fwrapv -O3 -Wall -Wstrict-prototypes -UNDEBUG -Isrc -I/usr/local/Cellar/python3/3.6.0/Frameworks/Python.framework/Versions/3.6/include/python3.6m -c build/temp.macosx-10.11-x86_64-3.6/_rtmixer.c -o build/temp.macosx-10.11-x86_64-3.6/build/temp.macosx-10.11-x86_64-3.6/_rtmixer.o
creating build/lib.macosx-10.11-x86_64-3.6
clang -bundle -undefined dynamic_lookup build/temp.macosx-10.11-x86_64-3.6/build/temp.macosx-10.11-x86_64-3.6/_rtmixer.o /Users/me/python-pa-ringbuffer/src/_pa_ringbuffer.abi3.so -o build/lib.macosx-10.11-x86_64-3.6/_rtmixer.abi3.so
ld: can't link with bundle (MH_BUNDLE) only dylibs (MH_DYLIB) file '/Users/me/python-pa-ringbuffer/src/_pa_ringbuffer.abi3.so' for architecture x86_64
clang: error: linker command failed with exit code 1 (use -v to see invocation)
error: command 'clang' failed with exit status 1
```

According to http://stackoverflow.com/a/24519922/ this seems to have to do with the `-bundle` and `-dynamiclib` options.

Any ideas how to make this work?

**Comment by tgarc** [comment]
Yeah I also tested on debian linux. Unfortunately I don't have any
experience developing for osx :/

**Comment by tgarc** [comment]
I tried this on python2.7 + Visual C++ for Python also no issues.

**Comment by mgeier** [comment]
@tgarc Thanks, that's great news!

I've written another message to the CFFI mailing list, I hope somebody will come up with a solution for macOS!

**Comment by tgarc** [comment]
That's annoying. Wish I could help, but I don't even have a mac. FWIW I also compiled with Python3.4 & Visual Studio 2010 and on my older redhat 6.7 setup at work without a hitch.

**Comment by mgeier** [comment]
@tgarc Thanks for the update, I hope somebody will come up with a solution for macOS!

BTW, you are talking about compiling ... I think you should also try to run it, because there might still be a dynamic link error at runtime?

**Comment by tgarc** [comment]
Yes I've run the code on all those systems, I'm using it in my own module.

**Comment by arigo** [comment]
@mgeier Sorry, still no feedback in the mailing list.  I don't have a Mac either; I only know that .so and .dylib files are somehow working differently from each other.  I *think* that by now, .dylib's are strictly more general than .so's, so that using .dylib everywhere should work.  But I don't know how that plays with the idea that CPython C extension modules should be .so's.

**Comment by mgeier** [comment]
@arigo Thanks for your answer!

TL;DR: I found a setting which seems to work on CPython/macOS, see the last few paragraphs of this comment.

I did some digging:

On macOS, there seem to be two types of "dynamic" libraries: ones for "loading" (`MH_BUNDLE`, with any extension, sometimes `.so`, sometimes `.dylib`, sometimes `.bundle`, sometimes something else) and ones for "linking" (`MH_DYLIB`, always with the extension `.dylib`, but sometimes also hidden in "frameworks", which are hopefully not relevant here).

"loading" seems to be not so picky and also work with `MH_DYLIB` nowadays, but "linking" still only works with `MH_DYLIB` and doesn't work with `MH_BUNDLE`. If I understood correctly, "loading" is what's done by `ffi.dlopen(...)`, while "linking" is what's done when I just specify the file name of the `.dylib` as command line option to the compiler.

The type of a given library can be found out with

```
otool -hv somelib.so
```

Here are some more details about the topic: http://docstore.mik.ua/orelly/unix3/mac/ch05_03.htm.

Some more random links:

http://stackoverflow.com/questions/2339679/what-are-the-differences-between-so-and-dylib-on-osx
http://stackoverflow.com/questions/32419594/how-to-create-a-dylib-c-extension-on-mac-os-x-with-distutils-and-or-setuptools

CPython seems to be expecting their extension modules to be `MH_BUNDLE`, see second comment here: https://github.com/python/cpython/blob/master/Python/dynload_next.c.
This was discussed in 2004: https://mail.python.org/pipermail/pythonmac-sig/2004-January/009925.html.

When I create a CFFI module with CPython, I indeed get a `MH_BUNDLE` (with the extension `.abi3.so`).

When I create a CFFI module with PyPy, I get a `MH_DYLIB` (with the extension `.pypy-41.so`).

When I try to compile this PR with CPython, I get the error message mentioned above: https://github.com/mgeier/python-rtmixer/pull/3#issuecomment-282786375.

I can compile it on PyPy without error message. However, I get an error on import:

```
>>>> import rtmixer
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/Users/me/python-rtmixer/src/rtmixer.py", line 7, in <module>
    from _rtmixer import ffi as _ffi, lib as _lib
ImportError: unable to load extension module '/Users/me/python-rtmixer/src/_rtmixer.pypy-41.so': dlopen(/Users/me/python-rtmixer/src/_rtmixer.pypy-41.so, 6): Library not loaded: build/lib.macosx-10.11-x86_64-2.7/_pa_ringbuffer.pypy-41.so
  Referenced from: /Users/me/python-rtmixer/src/_rtmixer.pypy-41.so
  Reason: image not found
```

The file `_pa_ringbuffer.pypy-41.so` (which is found successfully given its absolute path) seems to in turn link to the relative path `build/lib.macosx-10.11-x86_64-2.7/_pa_ringbuffer.pypy-41.so`. These two `.so` files are in fact identical (a copy was made during the CFFI build process), there just seems to be some mixup with the internally stored paths.

BTW, a little side note: `pypy` uses `cc` while CPython uses `clang` for compiling. On the Mac where I'm trying this, those are the same. I don't know if this means anything, I just wanted to mention it.

Now back to CPython:

When I manually replace the `-bundle` option with `-dynamiclib`, the compilation of `pa_ringbuffer` still works and I indeed get a `MH_DYLIB`. I can successfully import and use the `pa_ringbuffer` module, even though CPython expects a `MH_BUNDLE`.
I can also successfully compile the `rtmixer` module, but on import I get an error quite similar to the above:

```
>>> import rtmixer
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/Users/me/python-rtmixer/src/rtmixer.py", line 7, in <module>
    from _rtmixer import ffi as _ffi, lib as _lib
ImportError: dlopen(/Users/me/python-rtmixer/src/_rtmixer.abi3.so, 2): Library not loaded: build/lib.macosx-10.11-x86_64-3.6/_pa_ringbuffer.abi3.so
  Referenced from: /Users/me/python-rtmixer/src/_rtmixer.abi3.so
  Reason: image not found
```

I assume that problem could be somehow solved, but I don't know if it is the right approach, especially since CPython doesn't seem to "officially" support `MH_DYLIB` extension modules.

After all this playing around, I did one more random thing: I just removed the line

```
extra_objects=[_pa_ringbuffer.__file__],
```

... which previously added the `.dylib` file name to the compiler options. To my big suprise, now it works!
To be more exact: With CPython on macOS it seems to work if I remove the abovementioned line!
I don't know how and why, but it works ...

Does it have something to do with this line?

```
ffibuilder.include(_pa_ringbuffer_build.ffibuilder)
```

Does this do some magic on CPython/macOS that it doesn't do on PyPy/macOS and CPython/Linux?

**Comment by arigo** [comment]
I think I'll stay on my position that you're trying to do something that you shouldn't: (find a portable way to) import a compiled CPython extension module from another piece of C.  You would have the same troubles trying to do the same thing with regular CPython extension modules written in C, and you would have found the solution that fails to work on OS/X in the same way.

This problem is actually documented for CPython: https://docs.python.org/2/extending/extending.html#using-capsules ("capsules" are a custom but portable solution to that problem).

Maybe you should use a similar approach.  For example, cffi module A defines a (static) global structure ``a_api`` containing pointers to (static) functions, and expose it via ``cffi.def("""struct A_api a_api;""")``.  Then cffi module B has a (static) global function pointer to this structure, also exposed via ``cffi.def("""struct A_api *a_api_ptr;""")``.  Then you need one extra line of Python code after importing A and B, to link them together: ``b_ffi.a_api_ptr = ffi.addressof(a_ffi.a_api)``.

In this way, there is no need to compile B in such a way that it depends on A: the linking is done purely by that Python line.

See example at https://paste.pound-python.org/show/bNtITFyGgohi7TtgKcCV/

**Comment by mgeier** [comment]
Thanks a lot @arigo, that's extremely helpful, especially the working mini-example!

I've tried your suggestion, but with a little twist, since I didn't really like to force users to use the redirection via the API struct.
I thought if I can use a pointer to a struct, I might as well use multiple function pointers, and this way the user can call the functions just as in the original C API (without even knowing that now they are function pointers).

It works like a charm (for now I've tried it on CPython/Linux and PyPy/Linux)!

My first attempt is here: fbdd4257f7289477f1d8584706d9e308a5baacbb.
This only works with the corresponding changes in `pa_ringbuffer`: https://github.com/mgeier/python-pa-ringbuffer/pull/2/commits/19731b341cc095cddde4d9e1bd485a2e3f7a547d.
To simplify the usage, I've defined two strings `CDEF` and `SOURCE` that can be used in the "derived" module to avoid copying the C code.
I've also created a function `link_c_functions()` that encapsulates assigning stuff from one module to the other.
This already works, but I had also to add some casts, because in my Python code now I'm mixing the two CFFI instances. Other re-users will also have to do that.

To avoid all the casting and therefore simplifying the user code, I've tried to sneakily switch the CFFI instance from below the `RingBuffer` implementation.
This is done in 6dbcdda8af411fdd66e2f4bba9d899461e78bbca, which needs https://github.com/mgeier/python-pa-ringbuffer/pull/2/commits/d6b49b39b1b2ead807866196a645a68b5cf9373b.
I've renamed the function `link_c_functions()` to `link()`, because now it's doing more than that.

As far as I can tell, this works on Linux, I hope it will also work on other OSs.
But I'm not sure if I'm on the right track, probably there is a better solution ...

Is it a good idea to define and later assign all those function pointers?

Is it a good idea to pull a switcheroo on the CFFI instance of `pa_ringbuffer`?
I have the feeling that it is a bad idea, because if multiple modules "derive" from `pa_ringbuffer` and are used at the same time, each one will switch the CFFI instance ...
Probably I can make something that generates the `RingBuffer` class on demand, injecting the right CFFI instance?

Are there examples of other libraries that do a similar thing with CFFI?

**Comment by mgeier** [comment]
I've added another commit (df9b230c9c8a3e3951a44c438f2e1b6dcd75ed43) that uses a modified ring buffer class instead of monkey-patching the `pa_ringbuffer` module. This needs also an updated version of the `pa_ringbuffer` module: https://github.com/mgeier/python-pa-ringbuffer/commit/55595d8b71e36e8f9da23289ac739adb34b54f5e.

I think it should work now without problems, but I'm not sure if that's really the way it should be done ...

Any comments?

@tgarc I've seen that you are patching `pa_ringbuffer.py` in order to inject your own `ffi` and `lib` objects. What do you think about using something like the `link()` function (as in https://github.com/mgeier/python-pa-ringbuffer/commit/55595d8b71e36e8f9da23289ac739adb34b54f5e) instead?

Or is there probably something completely different that I could try to make re-using `pa_ringbuffer` easier?

BTW, the only project using `ffibuilder.include()` that I could find is https://github.com/gdanezis/bplib/blob/master/bplib/compile.py. I'm not sure, but I think they don't link to their "parent" CFFI module ...

**Comment by mgeier** [comment]
I've made another attempt for this in #5 (in combination with https://github.com/mgeier/python-pa-ringbuffer/pull/4).
This time, I completely removed the compilation part from `pa_ringbuffer` and only provide the Python code there. Plus, for convenience, the string that can be used in `ffibuilder.cdef()`.

The compilation now has to be done only once (in the `rtmixer` module) and there is only one CFFI instance involved, which hopefully makes things easier.

This might be a little more complicated to set up, but I hope it's still easy to use.
I've added some (hopefully accurate) documentation: https://github.com/mgeier/python-pa-ringbuffer/tree/foreign-init#usage.

It has the advantage that it also works by just copying `pa_ringbuffer.py` to some importable directory. Getting it from PyPI is optional (and currently it's not yet available there!).

@arigo Can you please have a quick look if that makes sense from a CFFI standpoint?

@tgarc Since you are the only user I know of, can you please check if that works for you?

Is #5 better than #3?

Any further suggestions?

**Comment by mgeier** [comment]
I'm closing this in favor of #5.
